### ISSUE TITLE
Can't cancelled invoice if  invoice state == STATE_PAID

### ISSUE THREAD

**Issue by omaxmo** [magento/magento2#18509]
﻿

```
vendor/magento/module-sales/Model/Order/Invoice.php
public function cancel()
    {
        if (!$this->canCancel()) {
            return $this;
        }
       ..

        if ($this->getState() == self::STATE_PAID) {
            $order->setTotalPaid($order->getTotalPaid() - $this->getGrandTotal());
            $order->setBaseTotalPaid($order->getBaseTotalPaid() - $this->getBaseGrandTotal());
        }
...
        return $this;
    }
```
I suppose that this code will never be used :
```
  if ($this->getState() == self::STATE_PAID) {
            $order->setTotalPaid($order->getTotalPaid() - $this->getGrandTotal());
            $order->setBaseTotalPaid($order->getBaseTotalPaid() - $this->getBaseGrandTotal());
        }

```
if we have:
```
  if (!$this->canCancel()) {
            return $this;
        }
```

Problem:
```
public function cancel()
    {
        if (!$this->canCancel()) {
            return $this;
        }

```
Possible solution: 
```
public function cancel()
    {
        if ($this->getState() == self::STATE_CANCELED) {
            return $this;
        }
```

### Preconditions

1. Tested on Magento 2.2.6

### Steps to reproduce

1. Load a invoice programmatically where state == STATE_PAID
2. Call $invoice->cancel()

### Expected result

1. The invoice should be canceled.

### Actual result

1. Can't cancel invoice if invoice state  ==  STATE_PAID

**Comment by magento-engcom-team** [comment]
Hi @omaxmo. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@omaxmo do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
Hi @SikailoISM. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by orlangur** [comment]
@engcom-backlog-nazar are you sure paid invoice is supposed to be cancellable?

**Comment by ghost** [comment]
Hi @omaxmo, @orlangur  you can't cancel invoice if state == paid, If state == paid, you can do the credit memo.

`When you create an invoice and realise that it needs to be cancelled, be aware that there are some legal restrictions involved. Technically, you should never delete an issued invoice but instead use a credit note to cancel the invoice.`

When you cancel an invoice:

1. Any invoiced lines on the order revert back to Not invoiced, however the quantity despatched is not changed.
2. If you've cancelled the invoice because the quantity despatched has been reduced for any lines, then you'll need to amend the despatch before reprinting the invoice.
3. If the order also needs to be cancelled, you'll need to remove the despatched quantities and cancel the order.
This is correct to **not cancel** invoice if state is paid.

**Comment by magento-engcom-team** [comment]
Hi @omaxmo. Thank you for your report. 
The issue has been fixed in magento/magento2#19462 by @omaxmo in 2.3-develop branch
Related commit(s):
 - [e3a8af60bd5cd7bc141da21bb57ec99d743f7db2](https://github.com/magento/magento2/commit/e3a8af60bd5cd7bc141da21bb57ec99d743f7db2)

The fix will be available with the upcoming 2.3.1 release.
